cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/MessagingMetricsCollectionService.java

```diff
@@ -89,10 +91,11 @@ protected void publish(Iterator<MetricValues> metrics) throws Exception {
       MetricValues metricValues = metrics.next();
       // Encode MetricValues into bytes
       recordWriter.encode(metricValues, encoder);
+      TopicPayload topicPayload = topicPayloads.get(Math.abs(metricValues.getTags().hashCode() % size));
       // Calculate the topic number with the hashcode of MetricValues' tags and store the encoded payload in the
       // corresponding list of the topic number
-      topicPayloads.get(Math.abs(metricValues.getTags().hashCode() % size))
-        .addPayload(encoderOutputStream.toByteArray());
+      topicPayload.addPayload(encoderOutputStream.toByteArray());
+      topicPayload.setMetricsTags(metricValues.getTags());
```

Comment:
The metrics context will be different for different MetricValues? How does this work?

Reply:
it would not change in most programs where there are just one metrics collector initialized in their program context, however certain containers like MR driver have more than one metrics collectors for (mapper/reducer) tasks. they would have multiple metric values where the context could be different.

currently we add the metric-values from the iterator and publish, we are using the last metricvalues tags from the iterator, should we rather use a <set> of tags and log that in the error ? am not sure whats the best approach here. would appreciate your suggestion.

Reply:
Maybe you can have a `Multimap<String, String>` to track context tag up to certain threshold.

```java
// Some arbitrary limit of number of unique context tag we are tracking for logging purpose
private static final int MAX_UNIQUE_CONTEXTS = 20;
private final Multimap<String, String> logContexts = LinkedHashMultimap.create();

void addPayload(Map<String, String> context, byte[] payload) {
  for (Map.Entry<String, String>entry = context.entrySet()) {
    if (logContexts.put(entry.getKey(), entry.getValue())) {
      int size = logContexts.get(entry.getKey()).size();
      if (size == MAX_UNIQUE_CONTEXTS - 1) {
        logContexts.put(entry.getKey(), "... and more values");
      }
    }
  }
  
  payloadSize += payload.length;
  payloads.add(payload);
}
```

Reply:
I am confused...